src/store/hdap.js

```diff
@@ -0,0 +1,402 @@
+        }))
+            .then(response => returnJson(response))
+    }
+
+    function getCommonHeaders(bodyString = null) {
+        const headers = {
+            'Accept': 'application/json',
+            'Content-Type': 'application/json'
+        }
+        if (isJwtValid()) headers['Authorization'] = `Bearer ${jwt.value}`
+        if (bodyString) headers['Content-Length'] = (new TextEncoder().encode(bodyString)).length
+        return headers
+    }
+
+    function getDateString(secondsFromNow) {
+        const future = Number(Math.floor(new Date().getTime() / 1000)) + Number(secondsFromNow)
+        return new Date(future * 1000).toISOString()
+    }
+
+    function getExpiredDateString() {
+        return new Date(0).toISOString() // '1970-01-01T00:00:00.000Z'
+    }
+
+    function getOptionalQueryString(params) {
+        return (params) ? '?' + new URLSearchParams(params) : ''
+    }
+
+    function isJwtValid() {
+        return new Date(expires.value).getTime() > new Date().getTime()
+    }
+
+    async function put(id, resource, additionalHeaders, optionalParams) {
+        if (!id || !resource) throw new Error('You must set the resource identifier and the resource')
+        const bodyString = JSON.stringify(resource)
+        const headers = getCommonHeaders(bodyString)
+        for (const [key, value] of Object.entries(additionalHeaders)) {
+            headers.headers[key] = value
+        }
+        return fetch(new Request(`${apiBase.value}/${id}${getOptionalQueryString(optionalParams)}`, {
+            method: 'PUT',
+            headers: headers,
+            body: bodyString
+        }))
+            .then(response => returnJson(response))
+    }
+
+    async function returnJson(response) {
+        if (!response.ok) throw new Error(`HTTP error with status: ${response.status}`)
+        return response.json()
+    }
+
+    async function throwError(message) {
+        message.value = message
+        throw new Error(message)
+    }
+
+    /*******************************
+     * Exported fields and functions
+     *******************************/
+
+    /** Base path to HDAP. */
+    const apiBase = ref('/hdap')
+
+    /** Resource of the currently authenticated user. */
+    const authenticatedUser = ref('')
+
+    /** Expiration date (string) of current bearer token. */
+    const expires = ref(getExpiredDateString())
+
+    /** The most human-readable name for the authenticated user. */
+    const friendlyUserName = ref('')
+
+    /** Bearer token for authenticated user. */
+    const jwt = ref('')
+
+    /** A human-readable message for the application user. */
+    const message = ref('')
+
+    /**
+     * Performs authentication and returns an access token.
+     *
+     * @param {*} id Resource identifier (_id) string, such as dc=com/dc=example/ou=People/uid=bjensen
+     * @param {*} passwd The password as a UTF-8 string
+     * @returns The access token.
+     */
+    async function authenticate(id, passwd) {
```

Comment:
I wonder if all these functions should actually be in a `src/helpers/hdap.js` file?

Trouble is, the whole thing is bound to authentication, which it would be nice to hide from the user of these APIs. There's a circular dependency if the functions depend on the authentication state and the store for authentication state depends on the helper authenticate function.

Reply:
On the other hand, the HDAP API should probably support anonymous, HTTP Basic, and HTTP Bearer auth.